### PR TITLE
feat: on chain route query

### PR DESCRIPTION
**Pull request by czarcas7ic**
Closes: #XXX

## What is the purpose of the change

Adds `osmosisd q poolmanager denom-pair-routes INPUT_COIN OUTPUT_DENOM`, which finds a basic route (min 1 hop, max 3 hops) to get to that denom. Some hard coded heurisitics:

1. CL liquidity has 1.5x more value than regular liquidity
2. If uosmo is the input or output denom, just search for a single hop

At time of upgrade, a route map is generated and store in a KVStore. After each epoch, this map is regenerated and restored. When comparing liquidity between routes to determine the best one, the routes are lexicographically ordered to make the iteration deterministic. This allows us to rely on this data structure in a state compatible manner.

## Testing and Verifying

This has been tested via CLI on a state exported testnet. Not only does it match the frontend routes, but at times takes better routes, and does so in a max of sub 1 second queries.

Some examples from this state exported testnet:
Screenshot 2023-11-21 at 4 43 08 PM

Gotest added for all new methods.

### CODE REVIEW

File: x/cosmwasmpool/genesis_test.go [hunk 1]
```diff
@@ -20,7 +20,7 @@ func (s *PoolModuleSuite) TestInitGenesis() {
 		expectedTotalLiquidity = expectedTotalLiquidity.Add(initalDefaultSupply...)
 	}
 
-	pools, err := s.App.CosmwasmPoolKeeper.GetPoolsWithWasmKeeper(s.Ctx)
```
Comment (czarcas7ic) [inline x/cosmwasmpool/genesis_test.go:23]: This method was removed in favor of utilizing GetPools to contain the keeper, and serializable method that does not contain the keeper

File: x/poolmanager/hooks.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+// At the end of each epoch, set the denom pair routes.
+func (k Keeper) AfterEpochEnd(ctx sdk.Context, epochIdentifier string, epochNumber int64) error {
+	_ = osmoutils.ApplyFuncIfNoError(ctx, func(cacheCtx sdk.Context) error {
+		_, err := k.SetDenomPairRoutes(cacheCtx)
+		return err
+	})
+
+	return nil
+}
```
Comment (czarcas7ic) [inline x/poolmanager/hooks.go:22]: We test that this properly sets the route store in TestGetDenomPairRoute

File: x/poolmanager/types/routes_test.go [hunk 3]
```diff
@@ -350,3 +350,66 @@ func TestIntermediateDenoms(t *testing.T) {
 		})
 	}
 }
+
+func TestAddEdge(t *testing.T) {
```
Comment (mattverse) [inline x/poolmanager/types/routes_test.go:354]: Can we add more test cases that tests:
1. start entry already exists, end entry does not
2. start entry and end entry already exists

and test that route is getting appended at the last index properly?

File: x/poolmanager/hooks.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+)
+
+func (k Keeper) BeforeEpochStart(ctx sdk.Context, epochIdentifier string, epochNumber int64) error {
+	return nil
+}
+
+// At the end of each epoch, set the denom pair routes.
+func (k Keeper) AfterEpochEnd(ctx sdk.Context, epochIdentifier string, epochNumber int64) error {
+	_ = osmoutils.ApplyFuncIfNoError(ctx, func(cacheCtx sdk.Context) error {
+		_, err := k.SetDenomPairRoutes(cacheCtx)
```
Comment (mattverse) [inline x/poolmanager/hooks.go:17]: Have we benchmarked this change? (Asking since we're adding additional computation on our epoch)

File: x/poolmanager/routes.go [hunk 5]
```diff
@@ -0,0 +1,569 @@
...
+							routeRoutes = append(routeRoutes, []*types.Route{startRoute, endRoute})
+						}
+					}
+				}
+			}
+		}
+	}
+
+	return routeRoutes
+}
```
Comment (mattverse) [inline x/poolmanager/routes.go:65]: nit: consider refactoring to following
```suggestion
func findTwoHopRoute(g types.RoutingGraphMap, start, end string) [][]*types.Route {
	var routeRoutes [][]*types.Route

	startRoutes, startExists := g.Graph[start]
	if !startExists {
		return routeRoutes
	}

	for token, routes := range startRoutes.InnerMap {
		tokenRoutes, tokenExists := g.Graph[token]
		if !tokenExists {
			continue
		}

		endRoutes, endExists := tokenRoutes.InnerMap[end]
		if !endExists {
			continue
		}

		for _, startRoute := range routes.Routes {
			startRoute.Token = token
			for _, endRoute := range endRoutes.Routes {
				endRoute.Token = end
				routeRoutes = append(routeRoutes, []*types.Route{startRoute, endRoute})
			}
		}
	}

	return routeRoutes
}

```

File: x/poolmanager/hooks.go (hunk 4 above)
Comment (czarcas7ic) [reply x/poolmanager/hooks.go:17]: The computation takes approx 1 second at maximum

File: x/poolmanager/routes.go [hunk 6]
```diff
@@ -0,0 +1,569 @@
...
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+
+	return routeRoutes
+}
```
Comment (mattverse) [inline x/poolmanager/routes.go:97]: nit: considering refactoring to following
```suggestion
func findThreeHopRoute(g types.RoutingGraphMap, start, end string) [][]*types.Route {
	var routeRoutes [][]*types.Route

	startRoutes, startExists := g.Graph[start]
	if !startExists {
		return routeRoutes
	}

	for token1, routes1 := range startRoutes.InnerMap {
		if token1 == start || token1 == end {
			continue
		}

		token1Routes, token1Exists := g.Graph[token1]
		if !token1Exists {
			continue
		}

		for token2, routes2 := range token1Routes.InnerMap {
			if token2 == start || token2 == end {
				continue
			}

			endRoutes, endExists := g.Graph[token2].InnerMap[end]
			if !endExists {
				continue
			}

			for _, startRoute := range routes1.Routes {
				clonedStartRoute := *startRoute
				clonedStartRoute.Token = token1

				for _, middleRoute := range routes2.Routes {
					clonedMiddleRoute := *middleRoute
					clonedMiddleRoute.Token = token2

					for _, endRoute := range endRoutes.Routes {
						clonedEndRoute := *endRoute
						clonedEndRoute.Token = end

						routeRoutes = append(routeRoutes, []*types.Route{&clonedStartRoute, &clonedMiddleRoute, &clonedEndRoute})
					}
				}
			}
		}
	}

	return routeRoutes
}

```

**Review by mattverse — COMMENTED** [review]
Will continue reviewing tomorrow

File: x/poolmanager/types/routes_test.go (hunk 3 above)
Comment (czarcas7ic) [reply x/poolmanager/types/routes_test.go:354]: Added further test cases here https://github.com/osmosis-labs/osmosis/pull/6895/commits/ebce499d64ae281dca157f27e3f267f7b5051408

File: x/poolmanager/routes.go [hunk 7]
```diff
@@ -0,0 +1,569 @@
...
+	directRouteCache = make(map[string]uint64)
+	spotPriceCache = make(map[string]osmomath.BigDec)
+}
+
+// findDirectRoute finds all direct routes between two tokens
+func findDirectRoute(g types.RoutingGraphMap, start, end string) []*types.Route {
+	if innerMap, exists := g.Graph[start]; exists {
+		if routes, exists := innerMap.InnerMap[end]; exists {
+			for i := range routes.Routes {
+				routes.Routes[i].Token = end
```
[the remark this answers is not in the document]
Comment (czarcas7ic) [reply x/poolmanager/routes.go:36]: So we return the end token in every instance here since this is the "out" token. Imagine we are swapping across three pools, and each of these pools has more than two tokens. If we just stored the pool route (2>5>8), its much harder for someone to extrapolate what each out token needs to be. Since we can just mark this at time of route determination, this allows us to give the exact route (pool ID and token out) for each hop.

TL;DR it just makes it simpler when estimating the swap via the route if we note the out token of each hop in the event it has more than two tokens.

[the remark this answers is not in the document]
Comment (czarcas7ic) [reply x/poolmanager/routes.go:36]: I think that was you question, lmk if that answers it

File: proto/osmosis/poolmanager/v1beta1/swap_route.proto [hunk 8]
```diff
@@ -38,3 +38,35 @@ message SwapAmountOutSplitRoute {
...
+}
+
+message Inner { repeated InnerMapEntry entries = 1; }
+
+message RoutingGraphEntry {
+  string key = 1;
+  Inner value = 2;
+}
+
+message RoutingGraph { repeated RoutingGraphEntry entries = 1; }
```
Comment (p0mvn) [inline proto/osmosis/poolmanager/v1beta1/swap_route.proto:72]: From discussions,I understand that this is to be able to serialize a map of maps.

However, I would not be able to tell otherwise. Could you please elaborate on this via comments?

File: x/cosmwasmpool/pool_module.go [hunk 9]
```diff
@@ -100,12 +100,15 @@ func (k Keeper) GetPool(ctx sdk.Context, poolId uint64) (poolmanagertypes.PoolI,
...
+			pool := model.CosmWasmPool{}
 			err := k.cdc.Unmarshal(value, &pool)
 			if err != nil {
 				return nil, err
 			}
-			return &pool, nil
+			return &model.Pool{
+				CosmWasmPool: pool,
+				WasmKeeper:   k.wasmKeeper,
+			}, nil
```
Comment (p0mvn) [inline x/cosmwasmpool/pool_module.go:111]: Could you please check whether this breaks the poolmanager `all-pools` query?

I vaguely remember that serialization issues as being the reason for not setting the wasm keeper here at the start

File: x/poolmanager/routes.go [hunk 10]
```diff
@@ -0,0 +1,597 @@
...
+
+	var twoHopPoolIDs [][]*types.Route
+	if inputDenom != OSMO && outputDenom != OSMO {
+		twoHopPoolIDs = findTwoHopRoute(routeMap, inputDenom, outputDenom)
+	}
+
+	var threeHopPoolIDs [][]*types.Route
+	if inputDenom != OSMO && outputDenom != OSMO {
+		threeHopPoolIDs = findThreeHopRoute(routeMap, inputDenom, outputDenom)
+	}
```
Comment (p0mvn) [inline x/poolmanager/routes.go:190]: Can this be done iteratively? What is the difference in handling each subsequent hop?

**Review by p0mvn — COMMENTED** [review]
Left some initial questions and comments. Nice work!

I would like to please request a sync walkthrough of this PR before I continue with the review.

Let's discuss it during our call tomorrow

File: x/cosmwasmpool/pool_module.go (hunk 9 above)
Comment (czarcas7ic) [reply x/cosmwasmpool/pool_module.go:111]: Confirming that this did not break all-pools. Here is an except from all-pools showing the CW pools:

```
- '@type': /osmosis.cosmwasmpool.v1beta1.CosmWasmPool
  code_id: "148"
  contract_address: osmo1gyg0pys40ex2f6a4dytd3ewpx2xfrsnt3rdc2t4j3s3jc9qx8kqsny066c
  instantiate_msg: eyJwb29sX2Fzc2V0X2Rlbm9tcyI6IFsiaWJjLzgyNDJBRDI0MDA4MDMyRTQ1N0QyRTEyRDQ2NTg4RkQzOUZCNTRGQjI5NjgwQzZDNzY2M0QyOTZCMzgzQzM3QzQiLCJpYmMvNEFCQkVGNEM4OTI2REREQjMyMEFFNTE4OENGRDYzMjY3QUJCQ0VGQzA1ODNFNEFFMDVENkU1QUEyNDAxRERBQiJdLCJhZG1pbiI6I>
  pool_id: "1211"
- '@type': /osmosis.cosmwasmpool.v1beta1.CosmWasmPool
  code_id: "148"
  contract_address: osmo10c8y69yylnlwrhu32ralf08ekladhfknfqrjsy9yqc9ml8mlxpqq2sttzk
  instantiate_msg: eyJwb29sX2Fzc2V0X2Rlbm9tcyI6IFsiaWJjL0QxODkzMzVDNkU0QTY4QjUxM0MxMEFCMjI3QkYxQzFEMzhDNzQ2NzY2Mjc4QkEzRUVCNEZCMTQxMjRGMUQ4NTgiLCJpYmMvNDk4QTA3NTFDNzk4QTBEOUEzODlBQTM2OTExMjNEQURBNTdEQUE0RkUxNjVENUM3NTg5NDUwNUI4NzZCQTZFNCJdLCJhZG1pbiI6I>
  pool_id: "1212"
```

File: proto/osmosis/poolmanager/v1beta1/swap_route.proto (hunk 8 above)
Comment (czarcas7ic) [reply proto/osmosis/poolmanager/v1beta1/swap_route.proto]: Further comments added here https://github.com/osmosis-labs/osmosis/pull/6895/commits/bbbc4396f9b3e6036a2fda02ecb89968cc2d2789

File: x/poolmanager/routes.go (hunk 10 above)
Comment (czarcas7ic) [reply x/poolmanager/routes.go]: I think this is what you were asking to be done https://github.com/osmosis-labs/osmosis/pull/6895/commits/9676acae390e1f68253fc8966eeafd0c8242115c

File: x/poolmanager/routes.go [hunk 11]
```diff
@@ -0,0 +1,537 @@
...
+			}
+			totalLiquidityInOsmo = totalLiquidityInOsmo.Add(liqInOsmo)
+		}
+		routeLiquidity[routeKey] = totalLiquidityInOsmo
+	}
+
+	// Extract and sort the keys from the routeLiquidity map
+	// This ensures deterministic selection of the best route
+	var keys []string
+	for k := range routeLiquidity {
```
Comment (mattverse) [inline x/poolmanager/routes.go:229]: Just curious, why not use routeliquidity Map directly sorted instead of creating a new string arr?

**Review by mattverse — COMMENTED** [review]
Went through code mostly lgtm! will do a final iteration after this one and then should be good to go

**Comment by mattverse** [comment]
@czarcas7ic Give me a ping once r4r!

File: proto/osmosis/poolmanager/v1beta1/swap_route.proto [hunk 12]
```diff
@@ -38,3 +38,34 @@ message SwapAmountOutSplitRoute {
...
+// retriving routes quickly and efficiently. Therefore, we use the repeated
+// message types for stateful structures and the map message types for
+// non-stateful structures to be used in retrieving routes.
+
+// Route message types to be used between both stateful and non-stateful
+// structures.
+message Route {
+  uint64 pool_id = 1;
+  string token = 2;
+}
```
Comment (p0mvn) [inline proto/osmosis/poolmanager/v1beta1/swap_route.proto:54]: Let's call this either `Pool` or `PoolInfo`

Route is a collection of `Pool`

**Review by p0mvn — COMMENTED** [review]
Left initial comments requesting more godocs where it would be helpful with the review
